src/main/java/org/cyclonedx/parsers/XmlParser.java

```diff
@@ -278,22 +289,56 @@ public boolean isValid(final InputStream inputStream, final Version schemaVersio
         return validate(inputStream, schemaVersion).isEmpty();
     }
 
-    private String identifySchemaVersion(final NodeList nodeList) {
-        for (int i=0; i<nodeList.getLength(); i++) {
-            final Node node = nodeList.item(i);
-            for (final Version version: Version.values()) {
-                if (version.getNamespace().equals(node.getNodeValue()))  {
-                    return version.getVersionString();
-                }
+    private String identifySchemaVersion(final InputSource in)
+        throws ParserConfigurationException, IOException, SAXException
+    {
+
+        List<String> namespaces = extractAllNamespaceDeclarations(in);
+
+        for (String namespaceUri : namespaces) {
+            String versionString = NAMESPACE_TO_VERSION_MAP.get(namespaceUri);
+            if (versionString != null) {
+                return versionString;
             }
         }
         return null;
     }
 
-    private NodeList extractAllNamespaceDeclarations(final InputSource in) throws XPathExpressionException {
-        final XPathFactory xPathFactory = XPathFactory.newInstance();
-        final XPath xPath = xPathFactory.newXPath();
-        final XPathExpression xPathExpression = xPath.compile("//namespace::*");
-        return (NodeList) xPathExpression.evaluate(in, XPathConstants.NODESET);
+    private List<String> extractAllNamespaceDeclarations(final InputSource in)
+        throws ParserConfigurationException, IOException, SAXException
+    {
+        Document doc = createSecureDocument(in);
```

Comment:
This is a good fix. However, I am not sure why a core library has to take a position to disable all use of DTD. What if an organization legitimately needs them especially for external references etc? Best to expose these as non-breaking settings and make it clear that users and applications are responsible for security and cannot offload security to core libraries.

Reply:
> This is a good fix. However, I am not sure why a core library has to take a position to disable all use of DTD. What if an organization legitimately needs them especially for external references etc? Best to expose these as non-breaking settings and make it clear that users and applications are responsible for security and cannot offload security to core libraries.

I thought about this but then I realized this part is not doing any parsing of the content, this is to get the schema version to parse to the right version, for the deserialization we use Jackson who by default has this protection enabled https://github.com/FasterXML/jackson-dataformat-xml/issues/190 so even if we wanted to do this we would need to change the default behavior of Jackson

Reply:
Disabling it is fine and desired IMO. An alternative would be to allow users to provide their own `DocumentBuilderFactory`, but defaulting to a locked-down version.

But that feels like something we can think about when there's an immediate need for it, i.e. when someone needs it.